nupic/research/frameworks/vernon/trainables.py

```diff
@@ -122,6 +93,132 @@ class ExtendedExpClass(RayActorHelperMethods, experiment_class):
 
         self._first_run = True
 
+    def _extend_experiment_class(self, experiment_class):
+        return experiment_class
+
+    def _train(self):
+        pass
+
+    def _should_stop(self):
+        """
+        Whether or not we should stop the experiment
+        """
+        # Check if we should stop the experiment
+        stop_status = self.procs[0].should_stop.remote()
+        if ray_utils.check_for_failure([stop_status]):
+            return ray.get(stop_status)
+        else:
+            # Stop on failures
+            return True
+
+    def _save(self, _=None):
+        self.logger.debug(f"_save: {self._trial_info.trial_name}({self.iteration})")
+        # All models are synchronized. Just save the state of first model
+        with warn_if_slow("ImagenetExperiment.get_state.remote"):
+            return ray.get(self.procs[0].get_state.remote())
+
+    def _restore(self, state):
+        # Restore the state to every process
+        state_id = ray.put(state)
+        ray.get([w.set_state.remote(state_id) for w in self.procs])
+
+        # Update current iteration using experiment epoch
+        self._iteration = ray.get(self.procs[0].get_current_epoch.remote())
+
+        self.logger.debug(f"_restore: {self._trial_info.trial_name}({self.iteration})")
+
+    def _create_workers(self, config):
+        pass
+
+    def _kill_workers(self):
+        for w in self.procs:
+            self.logger.warning(
+                f"Killing worker {w}, "
+                f"trial={self._trial_info.trial_name}({self.iteration})")
+            ray.kill(w)
+        self.procs = []
+
+    def _stop(self):
+        self.logger.debug(f"_stop: {self._trial_info.trial_name}({self.iteration})")
+        try:
+            status = [w.stop_experiment.remote() for w in self.procs]
+            # Wait until all remote workers stop
+            ray.get(status)
+            for w in self.procs:
+                w.__ray_terminate__.remote()
+            self.procs = []
+        except RayActorError as ex:
+            self.logger.warning("Failed to shutdown gracefully", exc_info=ex)
+            self._kill_workers()
+
+    def _process_config(self, config):
+        pass
+
+    def _run_iteration(self):
+        """Run one iteration of the experiment"""
+        status = []
+        for w in self.procs:
+            status.append(w.run_iteration.remote())
+
+        # Wait for remote functions and check for errors
+        if ray_utils.check_for_failure(status):
+            return ray.get(status)
+
+    def _process_result(self, result):
+        pass
+
+
+class RemoteProcessTrainable(RemoteTrainableBase):
```

Comment:
Is there an easy way to use this trainable with a `DataParallel` model? If anything, it would be useful for getting the benefit of multiple GPUs without the hassle of aggregating results. We can save this for later though when if the need ever arises (Our CL and MCL experiments use just one GPU at the moment, for instance.)

Reply:
For now, we can do that by extending `setup_experiment` (via a subclass or a mixin). Also, I'm considering removing the DataParallel usage from the Distributed* classes -- that doesn't seem like the right place for it.